File: basic_java_socket/src/main/java/org/example/ChatServer.java
package org.example;

import java.io.*;
import java.net.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

public class ChatServer {
    static int SIZE = 2;
    private List<ChatHandler> clients;
    private ExecutorService executor;
    public ChatServer() {
        clients = new ArrayList<>();
        executor = Executors.newFixedThreadPool(SIZE);
    }

    public void start(int port) {
        try (ServerSocket serverSocket = new ServerSocket(port)) {
            System.out.println("Chat Server started on port " + port);
            while (true) {

                Socket clientSocket = serverSocket.accept();
                if (clients.size()>=SIZE) {
                    System.out.println(clientSocket.getLocalAddress() + " Rejection : Chat Server is crowded ");
                    PrintWriter out = new PrintWriter(clientSocket.getOutputStream(), true);
                    out.println("Sorry, the chat server is currently crowded and unable to accommodate new connections.");
                    clientSocket.close();
                    continue;
                }

                System.out.println("New client connected: " + clientSocket.getInetAddress().getHostAddress());
                ChatHandler chatHandler = new ChatHandler(clientSocket);

                executor.execute(chatHandler);

            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void broadcast(String message) {
        for (ChatHandler client : clients) {
            try {
                client.writer.write(message + "\n");
                client.writer.flush();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    private void removeClient(ChatHandler client) {
        clients.remove(client);
        System.out.println("Client disconnected: " + client.getClientAddress());
    }

    private class ChatHandler extends Thread {
        private Socket clientSocket;
        private BufferedReader reader;
        private BufferedWriter writer;
        private String clientAddress;

        public ChatHandler(Socket clientSocket) {
            this.clientSocket = clientSocket;
            try {
                reader = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
                writer = new BufferedWriter(new OutputStreamWriter(clientSocket.getOutputStream()));
                clientAddress = clientSocket.getInetAddress().getHostAddress();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }

        public String getClientAddress() {
            return clientAddress;
        }

        public void run() {
            try {
                clients.add(this);
                String message;
                while ((message = reader.readLine()) != null) {
                    System.out.println("Received from " + clientAddress + ": " + message);
                    broadcast(message);
                }
            } catch (IOException e) {
                e.printStackTrace();
            } finally {
                try {
                    reader.close();
                    writer.close();
                    clientSocket.close();
                    removeClient(this);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    public static void main(String[] args) {
        ChatServer chatServer = new ChatServer();
        chatServer.start(12345);
    }
}

File: basic_kotlin_socket/src/main/kotlin/ChatServer.kt
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.PrintWriter
import java.net.ServerSocket

val clients = mutableListOf<PrintWriter>()

fun main() = runBlocking {
    val server = ServerSocket(1234)
    println("채팅 서버가 시작되었습니다.")

    while (true) {
        val clientSocket = server.accept()
        println("클라이언트가 연결되었습니다.")
        val reader = BufferedReader(InputStreamReader(clientSocket.getInputStream()))
        val writer = PrintWriter(clientSocket.getOutputStream(), true)

        clients.add(writer)

        GlobalScope.launch(Dispatchers.IO) {
            while (true) {
                val message = reader.readLine()
                if (message != null) {
                    println("클라이언트로부터 받은 메시지: $message")
                    broadcast(message)
                }
            }
        }
    }
}

fun broadcast(message: String) {
    for (client in clients) {
        client.println(message)
    }
}
